src/cpp/fastdds/publisher/DataWriterImpl.cpp

```diff
@@ -568,7 +648,13 @@ void DataWriterImpl::InnerDataWriterListener::onWriterChangeReceivedByAll(
         RTPSWriter* /*writer*/,
         CacheChange_t* ch)
 {
-    if (data_writer_->qos_.durability().kind == VOLATILE_DURABILITY_QOS)
+    if (data_writer_->type_->m_isGetKeyDefined &&
+            (NOT_ALIVE_UNREGISTERED == ch->kind ||
+            NOT_ALIVE_DISPOSED_UNREGISTERED == ch->kind))
```

Comment:
Should we also handle `NOT_ALIVE_DISPOSED`?
```suggestion
    if (data_writer_->type_->m_isGetKeyDefined && (ALIVE != ch->kind))
```

Reply:
When an instance is disposed, the resources haven't to be released. Only when unregistering the instance.